pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -185,30 +145,46 @@ func (configgen *ConfigGeneratorImpl) buildClusters(proxy *model.Proxy, req *mod
 	}
 	resources = cb.normalizeClusters(resources)
 
-	if cacheStats.empty() {
-		return resources, model.DefaultXdsLogDetails
+	built := sets.New()
+	for _, c := range resources {
+		deleted.Delete(c.Name)
+		built.Insert(c.Name)
 	}
-	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
-}
 
-func shouldUseDelta(updates *model.PushRequest) bool {
-	return updates != nil && deltaAwareConfigTypes(updates.ConfigsUpdated) && len(updates.ConfigsUpdated) > 0
-}
-
-// deltaAwareConfigTypes returns true if all updated configs are delta enabled.
-func deltaAwareConfigTypes(cfgs map[model.ConfigKey]struct{}) bool {
-	for k := range cfgs {
-		if !deltaConfigTypes.Contains(k.Kind.Kind) {
-			return false
-		}
+	if alreadyKnown != nil {
+		log.Errorf("after build\n~:%v\n-:%v\n+:%v", sets.New(alreadyKnown...).SortedList(), deleted.SortedList(), built.SortedList())
 	}
-	return true
+	if cacheStats.empty() {
+		return resources, deleted.SortedList(), model.DefaultXdsLogDetails
+	}
+	return resources, deleted.SortedList(), model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
 // buildOutboundClusters generates all outbound (including subsets) clusters for a given proxy.
-func (configgen *ConfigGeneratorImpl) buildOutboundClusters(cb *ClusterBuilder, proxy *model.Proxy, cp clusterPatcher,
-	services []*model.Service,
+func (configgen *ConfigGeneratorImpl) buildOutboundClusters(cb *ClusterBuilder, proxy *model.Proxy,
+	cp clusterPatcher, services []*model.Service, updated map[model.ConfigKey]struct{}, deleted sets.Set, delta bool,
 ) ([]*discovery.Resource, cacheStats) {
+	// make global-ish so can modify outside
+	var oldCacheKeys map[string]model.XdsCacheEntry
+	if proxy.WatchedResources[v3.ClusterType] == nil {
+		// placeholder;this will force generation because there are no cache keys
+		oldCacheKeys = make(map[string]model.XdsCacheEntry)
+	} else {
+		oldCacheKeys = proxy.WatchedResources[v3.ClusterType].CacheKeys
+	}
+	// nil -> delete all/reset oldCacheKeys
+	if configgen.InvalidatedResources == nil {
+		oldCacheKeys = map[string]model.XdsCacheEntry{}
+	}
+	// nonzero length -> delete some
+	if len(configgen.InvalidatedResources) != 0 {
+		// remove from oldCacheKeys
+		for _, k := range configgen.InvalidatedResources {
+			delete(oldCacheKeys, k)
+		}
+	}
+	// set to length 0 for next iteration (not nil)
+	configgen.InvalidatedResources = []string{}
```

Comment:
This is a global resource, but we are clearing it on a single call for a single proxy.

I think the write has the same issue